torch/optim/_functional.py

```diff
@@ -357,3 +357,51 @@ def asgd(params: List[Tensor],
             ax.add_(param.sub(ax).mul(mu))
         else:
             ax.copy_(param)
+
+
+def radam(params: List[Tensor],
+          grads: List[Tensor],
+          exp_avgs: List[Tensor],
+          exp_avg_sqs: List[Tensor],
+          state_steps: List[int],
+          *,
+          beta1: float,
+          beta2: float,
+          lr: float,
+          weight_decay: float,
+          eps: float):
+    r"""Functional API that performs RAdam algorithm computation.
+
+    See :class:`~torch.optim.Adam` for details.
+    """
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        step = state_steps[i]
+
+        bias_correction1 = 1 - beta1 ** step
+        bias_correction2 = 1 - beta2 ** step
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # Decay the first and second moment running average coefficient
+        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+        exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+        # maximum length of the approximated SMA
+        rho_inf = 2 / (1 - beta2) - 1
+        # compute the length of the approximated SMA
+        rho_t = rho_inf - 2 * step * (beta2 ** step) / bias_correction2
+
+        if rho_t > 5.:
+            # Compute the variance rectification term and update parameters accordingly
+            rect = math.sqrt((rho_t - 4) * (rho_t - 2) * rho_inf / ((rho_inf - 4) * (rho_inf - 2) * rho_t))
+            denom = (exp_avg_sq.sqrt() / math.sqrt(bias_correction2))
+            step_size = lr * rect / bias_correction1
+            param.addcdiv_(exp_avg, denom, value=-step_size)
+        else:
+            step_size = lr / bias_correction1
```

Comment:
based on [algorithm 2](https://arxiv.org/pdf/1908.03265.pdf), shouldn't `lr` be momentum here?

Reply:
the only difference between the two case should be `r_t l_t`. can you write the code in a way that makes this more explicit?

Reply:
Yes, i agree that there is a confusion with the variable names here and showing the difference is only r_t l_t would be much better presentation.

Therefore I adjusted the code accordingly. (my apologies for previous confusion, actually lr here is step-size in the paper, but now code should be much more readable)